### PR TITLE
Add TypeScript guide

### LINKED ISSUE #6: Add TypeScript Guide
Cover atleast
- [ ] Basic Types
- [ ] Control Flow narrowing
- [ ] Extending JSX
- [ ] Mention ! in refs break
- [ ] Typing Context by ReturnType

### PR DESCRIPTION
**Pull request by edemaine**
Fixes #6.

* [x] `.tsconfig`, `@jsxImportSource`
* [x] Basic Types (though could imagine covering more)
* [x] Control Flow narrowing
* [x] Extending JSX for custom Events, Properties, Attributes, and Directives
* [x] `!` in refs
* [x] Typing Context by `ReturnType` or default value

### CODE REVIEW

File: langs/en/guides/typescript.md [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+A common pattern is to use
+[`<Show>`](https://www.solidjs.com/docs/latest/api#%3Cshow%3E)
+to display data only when that data is defined:
+
+```ts
+const [name, setName] = createSignal<string>();
+
+return (
+  <Show when={name()}>
+    GREETINGS {name().toUpperCase()}!
```
Comment (davedbase) [inline langs/en/guides/typescript.md]: Minor suggestion not to use caps lock. Nothing really wrong with it aside from being a bit awkward. Personal preference :-)

Comment (edemaine) [reply langs/en/guides/typescript.md]: I wanted to use a `String` method here so it's clear that this code won't work when `name()` is `undefined` (and TypeScript will complain as a result). Any suggestions for a replacement?

Update: I replaced the example with `.replace(/\s+/, '\xa0')` (make whitespace nonbreaking)

**Comment by Jutanium** [comment]
Thanks so much for this!
I apparently didn't use GitHub quite right and instead added my readability changes to a [separate branch.](https://github.com/solidjs/solid-docs/tree/edemaine-typescript-guide)

One question about `refs`: I'm not sure that it's necessary to apply the non-null assertion while using the ref attribute , e.g.`<div ref={divRef!}>`. Intellisense is a bit annoyed if we don't, but it arguably misrepresents what's happening (like you say in the guide). I might have missed a conversation about this.

**Comment by Jutanium** [comment]
Following the discussion on Discord, I [revised the refs section](https://github.com/solidjs/solid-docs/commit/bc4f591834d583c81fc4c291cbb7f92bb96fb4d6). As far as I'm concerned, we're good to go!

**Comment by edemaine** [comment]
Thanks @Jutanium for the great edit!  I have a few comments/questions:

* ~~Do you think it's worth keeping a definition of the `Signal<T>` type? I admit it's not that common to use, but I personally liked knowing about it... But maybe this is better suited for the API doc? (It's not there yet.)~~ Oops I thought you removed this but you just moved it.
* As discussed further in Discord, I think it's better to include the `ref={ref!}` version. Compare these:

  `!` on `ref` correctly detects that the ref isn't initialized until the JSX block:

  VSCode screenshot

  `!` on `let` hides any such errors:

  VSCode screenshot

  The first form works with `onMount` and `createEffect` too (though it's a little odd that it does, as it works with `createRenderEffect` and `createMemo` even though it shouldn't):

  VSCode screenshot

  So I'd argue for reverting this section to the old version, possibly with a note about being careful with `createMemo`/`createRenderEffect`.

* I realize I accidentally deleted the `!` in the first fix to `Show`. One of us should fix this.

**Pull request merged**